exercise/submission_models.py

```diff
@@ -581,11 +586,18 @@ def set_ready(self, approve_unofficial=False):
                 "site": settings.BASE_URL,
             })
 
+        if not PendingSubmissionManager.is_grader_stable:
```

Comment:
I must have originally missed this. I think this line is very broken currently. Without the parentheses `()`, `is_grader_stable` refers to the function object, which I assume is **always truthy** in the `if` condition.

In addition, manager instances are used via the `objects` field of the model (or other field similarly, but `objects` is the de facto standard name for the default manager). The method `is_grader_stable` has not been marked as a static method even though you are now trying to use it as if it were static.

```python
# normal method, not static
if not PendingSubmission.objects.is_grader_stable():

# if the method were static, Python still allows calling it via the object instance
if not PendingSubmission.objects.is_grader_stable():

# but static methods could also be called like this.
# If we used this case, then it was not very useful to move
# the method definition to the PendingSubmissionManager class.
if not PendingSubmissionManager.is_grader_stable():
```

Reply:
In the method definition, note that normal methods MUST have at least one parameter `self` (the name does not technically matter, but it would be a dire breach of the best practices to use any other name than `self`).

If we used static methods, they would need the decorator `@staticmethod`. No `self` parameter is used in a static method.
https://docs.python.org/3/library/functions.html#staticmethod

Manager objects can (should even?) refer to the query methods via `self`. It is not necessary to hardcode the partner model class's name `PendingSubmission` when you can use `self`.

NEW SOLUTION
```python
class PendingSubmissionManager(models.Manager):
    def is_grader_stable(self):
        total_retries = self.aggregate(sum=models.Sum('num_retries'))['sum']
        return not (total_retries and total_retries > settings.GRADER_STABLE_THRESHOLD)
```
https://docs.djangoproject.com/en/3.2/topics/db/managers/

OLD broken version, currently in the master branch. No `self` parameter nor `@staticmethod`.
```python
class PendingSubmissionManager(models.Manager):
    def is_grader_stable():
        total_retries = PendingSubmission.objects.aggregate(sum=models.Sum('num_retries'))['sum']
        return not (total_retries and total_retries > settings.GRADER_STABLE_THRESHOLD)
```

Reply:
Fixed in https://github.com/apluslms/a-plus/pull/1103